### DISCUSSION TITLE
Push Vs Pull

### DISCUSSION THREAD

**Discussion by IvanMerrill** [autometrics-dev/autometrics-shared#40] [Ideas]
Currently the metrics are exposed via HTTP to be scraped, which is the Prometheus model. However the OTEL SDKs also support pushing metrics to an endpoint, commonly the OTEL collector, which does have advantages in certain scenarios such as in secure networks. In [discussion autometrics-dev/autometrics-shared#32](https://github.com/orgs/autometrics-dev/discussions/7) there's talk of creating a default option for creating the http endpoint, which I agree with, but I wonder about push support?

One option is to just use the OTEL collector to scrape the endpoint and send data wherever, which is a perfectly reasonable solution

**Comment by emschwartz** [comment]
I think this is an interesting thing to consider.

The main challenge I see with doing anything that isn't Prometheus-related is that right now we're writing PromQL queries to help you understand the data, and I think that's at least half the value of Autometrics. If we were going to push data somewhere else, we'd probably want to generate some other kinds of queries for you, but that brings up the question of what other query languages would be worth supporting.

That said, pushing metrics is definitely useful to explore for the Typescript implementation to support client-side metrics.

Maybe for the moment it would be most useful to add an example to each repo that shows how you would use the OTel library to push the metrics to the collector?

**Reply by egtwp** [reply]
Hello there, I hope an external point of view is helpful.
I appreciate your concerns about supporting other query languages. 
However, in my opinion, these concerns are not strictly correlated because someone could just go to Prometheus using a different collecting mechanism/architecture. 

I also really appreciate the idea of adding an example of a OTEL push mechanism at your repos. We’re providing an example of that in RUST in this [repository](https://github.com/egtwp/rs-autometrics-otlp-exporter) but it’s not fully working as described in this [issue](https://github.com/autometrics-dev/autometrics-rs/issues/91).

**Reply by emschwartz** [reply]
I think what @egtwp added to the Rust repo may be sufficient for now. Documenting and providing examples of how to use the push capabilities seems useful, but we may not need to add specific functionality to support this in each autometrics library.

The one exception I'd make to this is the Typescript repo, because in order for that to be used on the frontend (the main use case for JS/TS in the first place), it needs to use push. In that case, it's probably worth having specific functionality to make it easy to push the metrics somewhere.
